### PR TITLE
[Draft] Update HGCAL pset for Phase-2 HLT menu

### PR DESCRIPTION
**Pull request by srimanob**
#### PR description:
This PR updates the pset used in Phase-2 HGCAL HLT. The update follows the pset defined in
https://github.com/cms-sw/cmssw/blob/master/SimCalorimetry/HGCalSimProducers/python/hgcalDigitizer_cfi.py

To complete on PR test, and example of benefit from the new workflow, this PR also introduces 
- a new workfow `.76` which the `DIGI` and `HLT:@relval2026` steps are running together instead of `HLT:@fake`.  The new workflow will allow us to develop Phase2 HLT DQM and sequence to be used in future Phase-2 production. This PR is an effort to help on validation as mentioned in https://github.com/cms-sw/cmssw/issues/39362.
- a minimal change to have Validation HLT in sequence. There are still some warning massages, this will be cleaned up.

#### PR validation:
Run test with updated `39434.76`.

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:
None.

### CODE REVIEW

**Comment by srimanob** [comment]
@cmsbuild please test

**Comment by srimanob** [comment]
The local test (based on CMSSW_12_6_0_pre2) seems to work fine (I use 39434.75, then modify DIGI-HLT step.) It shows some error message, but did not crash.

Error messages that I got, but nothing about `l1verticesEmulation` as in PR test.
```
%MSG-e TriggerSummaryProducerAOD:   TriggerSummaryProducerAOD:hltTriggerSummaryAOD 19-Sep-2022 20:39:56 CEST  Run: 1 Event: 9
Uunknown pid: 2:813 FilterTag / Key: L1TkEleDouble12Filter::HLT / 0of1 CollectionTag / Key: l1ctLayer1EG:L1TkEleEE:HLT / 0 CollectionType: St6vectorIN3l1t10TkElectronESaIS1_EE
%MSG
%MSG-e TriggerSummaryProducerAOD:   TriggerSummaryProducerAOD:hltTriggerSummaryAOD 19-Sep-2022 20:39:56 CEST  Run: 1 Event: 9
Uunknown pid: 2:813 FilterTag / Key: L1TkEleSingle25Filter::HLT / 0of1 CollectionTag / Key: l1ctLayer1EG:L1TkEleEE:HLT / 0 CollectionType: St6vectorIN3l1t10TkElectronESaIS1_EE
%MSG
%MSG-e TriggerSummaryProducerAOD:   TriggerSummaryProducerAOD:hltTriggerSummaryAOD 19-Sep-2022 20:39:56 CEST  Run: 1 Event: 9
Uunknown pid: 2:813 FilterTag / Key: L1TkEleSingle36Filter::HLT / 0of1 CollectionTag / Key: l1ctLayer1EG:L1TkEleEE:HLT / 0 CollectionType: St6vectorIN3l1t10TkElectronESaIS1_EE
%MSG
%MSG-e TriggerSummaryProducerAOD:   TriggerSummaryProducerAOD:hltTriggerSummaryAOD 19-Sep-2022 20:39:56 CEST  Run: 1 Event: 9
Uunknown pid: 2:823 FilterTag / Key: L1TkEmDouble24Filter::HLT / 0of2 CollectionTag / Key: l1ctLayer1EG:L1TkEmEB:HLT / 0 CollectionType: St6vectorIN3l1t4TkEmESaIS1_EE
%MSG
%MSG-e TriggerSummaryProducerAOD:   TriggerSummaryProducerAOD:hltTriggerSummaryAOD 19-Sep-2022 20:39:56 CEST  Run: 1 Event: 9
Uunknown pid: 2:824 FilterTag / Key: L1TkEmDouble24Filter::HLT / 1of2 CollectionTag / Key: l1ctLayer1EG:L1TkEmEE:HLT / 0 CollectionType: St6vectorIN3l1t4TkEmESaIS1_EE
%MSG
%MSG-e TriggerSummaryProducerAOD:   TriggerSummaryProducerAOD:hltTriggerSummaryAOD 19-Sep-2022 20:39:56 CEST  Run: 1 Event: 9
Uunknown pid: 2:823 FilterTag / Key: L1TkEmSingle37Filter::HLT / 0of2 CollectionTag / Key: l1ctLayer1EG:L1TkEmEB:HLT / 0 CollectionType: St6vectorIN3l1t4TkEmESaIS1_EE
%MSG
%MSG-e TriggerSummaryProducerAOD:   TriggerSummaryProducerAOD:hltTriggerSummaryAOD 19-Sep-2022 20:39:56 CEST  Run: 1 Event: 9
Uunknown pid: 2:824 FilterTag / Key: L1TkEmSingle37Filter::HLT / 1of2 CollectionTag / Key: l1ctLayer1EG:L1TkEmEE:HLT / 0 CollectionType: St6vectorIN3l1t4TkEmESaIS1_EE
%MSG
%MSG-e TriggerSummaryProducerAOD:   TriggerSummaryProducerAOD:hltTriggerSummaryAOD 19-Sep-2022 20:39:56 CEST  Run: 1 Event: 9
Uunknown pid: 2:823 FilterTag / Key: L1TkEmSingle51Filter::HLT / 0of1 CollectionTag / Key: l1ctLayer1EG:L1TkEmEB:HLT / 0 CollectionType: St6vectorIN3l1t4TkEmESaIS1_EE
%MSG
%MSG-e TriggerSummaryProducerAOD:   TriggerSummaryProducerAOD:hltTriggerSummaryAOD 19-Sep-2022 20:39:56 CEST  Run: 1 Event: 9
Uunknown pid: 2:823 FilterTag / Key: L1TkIsoEmDouble12Filter::HLT / 0of1 CollectionTag / Key: l1ctLayer1EG:L1TkEmEB:HLT / 0 CollectionType: St6vectorIN3l1t4TkEmESaIS1_EE
%MSG
%MSG-e TriggerSummaryProducerAOD:   TriggerSummaryProducerAOD:hltTriggerSummaryAOD 19-Sep-2022 20:39:56 CEST  Run: 1 Event: 9
Uunknown pid: 2:823 FilterTag / Key: L1TkIsoEmSingle22Filter::HLT / 0of1 CollectionTag / Key: l1ctLayer1EG:L1TkEmEB:HLT / 0 CollectionType: St6vectorIN3l1t4TkEmESaIS1_EE
%MSG
%MSG-e TriggerSummaryProducerAOD:   TriggerSummaryProducerAOD:hltTriggerSummaryAOD 19-Sep-2022 20:39:56 CEST  Run: 1 Event: 9
Uunknown pid: 2:823 FilterTag / Key: L1TkIsoEmSingle36Filter::HLT / 0of1 CollectionTag / Key: l1ctLayer1EG:L1TkEmEB:HLT / 0 CollectionType: St6vectorIN3l1t4TkEmESaIS1_EE
%MSG

```

**Comment by Martin-Grunewald** [comment]
Can we disentangle the update of the menu itself from the update of @relval2026? I believe the latter needs much more work as currently the Phase-2 75e33 menu runs in a very special workflow only.
These errors: ```%MSG-e TriggerSummaryProducerAOD``` also point to a misconfiguration...

**Comment by srimanob** [comment]
hold

This PR can hold until the L1T transition is done. Thing should be more clear.

**Comment by srimanob** [comment]
By the way, the issue seems to come between 12_6_0_pre2 and last IB (CMSSW_12_6_X_2022-09-19-2300). Testing this code works with 12_6_0_pre2. Strange that we don't spot it with wf `39434.75` in IB test.

Any idea @cms-sw/l1-l2 ? What are just merged and may cause this?

**Comment by cecilecaillol** [comment]
@srimanob Between the two releases you mention, our module renaming PR (https://github.com/cms-sw/cmssw/pull/39244) has been merged. l1ct* objects are now called l1t*, maybe this has not been propagated everywhere to the HLT code, and was not tested in the PR automatic tests

**Comment by srimanob** [comment]
Hi @Martin-Grunewald @cecilecaillol 

The issue in PR test seems to happen only when running step-2 (DIGI-L1Track-L1-HLT@relval2026). If `HLT@relval2026` is reruning, i.e. in 39434.75 wf, the issue will not happen. This answers why we don't spot any issue in the IB test. This make me worry on what exactly we are running in .75 wf, and why is the behavior of HLT different. I assume it should give the same result, not depend on when it will run.

**Comment by srimanob** [comment]
@cmsbuild please test

File: HLTrigger/Configuration/python/HLT_75e33/modules/siPhase2Clusters_cfi.py [hunk 1]
```diff
@@ -5,3 +5,6 @@
     maxNumberClusters = cms.uint32(0),
     src = cms.InputTag("mix","Tracker")
 )
+
+from Configuration.ProcessModifiers.premix_stage2_cff import premix_stage2
+premix_stage2.toModify(siPhase2Clusters, src = "mixData:Tracker")
```
Comment (missirol) [inline HLTrigger/Configuration/python/HLT_75e33/modules/siPhase2Clusters_cfi.py:10]: The following
```py
import FWCore.ParameterSet.Config as cms

from RecoLocalTracker.SiPhase2Clusterizer.phase2TrackerClusterizer_cfi import siPhase2Clusters as _siPhase2Clusters
siPhase2Clusters = _siPhase2Clusters.clone()
```
should be equivalent, and would be easier to maintain, and less error-prone.

**Comment by missirol** [comment]
> Can we disentangle the update of the menu itself from the update of @relval2026?

@srimanob 

Thanks for trying to make this Phase-2 wfs work, but I would suggest to also follow Martin's suggestion. Right now, this PR contains different unrelated updates.

**Comment by srimanob** [comment]
Hi @missirol @Martin-Grunewald 

When you said to disentangle the update of the menu from the update of @relval2026, do you mean open another PR for 
`Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py`
and changes the phase-2 menu from `fake2` to `relval2026` ?
I can do that, but it seems the change is very minimal for this case. Only changing of a menu of baseline Phase-2 geometry.
This PR without the change of menu can't spot anything. So we have to trigger the test together anyways.

However, I can make in this way if you think it is more appropriate. Thx.

File: HLTrigger/Configuration/python/HLT_75e33/modules/siPhase2Clusters_cfi.py (hunk 1 above)
Comment (srimanob) [reply HLTrigger/Configuration/python/HLT_75e33/modules/siPhase2Clusters_cfi.py:10]: Sure, I can do that. Thanks for suggestion.

I tried to make very minimal change as much as possible without touching original code. I assume we have hard-copy coding around for 75e33 menu. I will fix it together with configuration when this comment, https://github.com/cms-sw/cmssw/pull/39450#issuecomment-1252829025, is confirmed.

**Comment by missirol** [comment]
Hi @srimanob ,

in practise, I was suggesting to separate the HGCal updates from the rest of this PR, if possible.

Could you please clarify where these HGCal come from, and if they change physics outputs in any way?

**Comment by missirol** [comment]
> This make me worry on what exactly we are running in .75 wf, and why is the behavior of HLT different. I assume it should give the same result, not depend on when it will run.

The underlying issue is the wrong `InputTag` in the L1T module `l1tHPSPFTauProducerPF` (as you spotted, and already fixed). The reason this did not happen in `39434.75` is that that module comes from the `L1` step, which is only part of `step2`, and not `step4`. In `step2`, we had the Fake2 HLT menu with [1], so `TriggerSummaryProducerAOD` was not calling `l1tHPSPFTauProducerPF`. With this PR, `step2` includes `HLT:@relval2026` with [2], which called the misconfigured `l1tHPSPFTauProducerPF`.

The reason why `step2` did not crash before because of this misconfigured L1T module (regardless of HLT) is likely that `l1tHPSPFTauProducerPF` does not actually run, because nothing consumes it (not another producer, nor an output module).

Looks like L1T is trying to fix this in https://github.com/cms-sw/cmssw/pull/39401#discussion_r976148112.

[1]
```py
hltTriggerSummaryAOD = cms.EDProducer("TriggerSummaryProducerAOD",
    moduleLabelPatternsToMatch = cms.vstring('hlt*'),
    moduleLabelPatternsToSkip = cms.vstring(),
    processName = cms.string('@'),
    throw = cms.bool(False)
)
```

[2]
```py
process.hltTriggerSummaryAOD = cms.EDProducer("TriggerSummaryProducerAOD",
    moduleLabelPatternsToMatch = cms.vstring(
        'hlt*',
        'l1t*'
    ),
    moduleLabelPatternsToSkip = cms.vstring(),
    processName = cms.string('*'),
    throw = cms.bool(False)
)
```

**Comment by srimanob** [comment]
> in practise, I was suggesting to separate the HGCal updates from the rest of this PR, if possible.
> Could you please clarify where these HGCal come from, and if they change physics outputs in any way?

Hi @missirol 
The configuration of HGCAL comes from
https://github.com/cms-sw/cmssw/blob/master/SimCalorimetry/HGCalSimProducers/python/hgcalDigitizer_cfi.py

If you do the config with `HLT:@fake2`, i.e. `DIGI:pdigi_valid,L1TrackTrigger,L1,DIGI2RAW,HLT:@fake2`, you will get 
```
process.HGCAL_noise_fC = cms.PSet(
    doseMap = cms.string(''),
    scaleByDose = cms.bool(False),
    scaleByDoseAlgo = cms.uint32(0),
    scaleByDoseFactor = cms.double(1),
    values = cms.vdouble(0.32041011999999996, 0.384492144, 0.32041011999999996)
)
process.HGCAL_noise_heback = cms.PSet(
    doseMap = cms.string(''),
    noise_MIP = cms.double(0.01),
    referenceIdark = cms.double(-1),
    referenceXtalk = cms.double(-1),
    scaleByDose = cms.bool(False),
    scaleByDoseAlgo = cms.uint32(0),
    scaleByDoseFactor = cms.double(1),
    sipmMap = cms.string('')
)
```

But if you do `DIGI:pdigi_valid,L1TrackTrigger,L1,DIGI2RAW,HLT:@relval2026` as in the release now, it seems what defined in 
- https://github.com/cms-sw/cmssw/blob/master/HLTrigger/Configuration/python/HLT_75e33/psets/HGCAL_noise_heback_cfi.py
- https://github.com/cms-sw/cmssw/blob/master/HLTrigger/Configuration/python/HLT_75e33/psets/HGCAL_noise_fC_cfi.py
will overwrite `process.HGCAL_noise_fC`, `process.HGCAL_noise_heback`. This is not good for mixing process.

process.HGCAL_noise_fC = cms.PSet(
    doseMap = cms.string(''),
    scaleByDose = cms.bool(False),
    scaleByDoseAlgo = cms.uint32(0),
    values = cms.vdouble(0.32041012, 0.384492144, 0.32041012)
)
process.HGCAL_noise_heback = cms.PSet(
    doseMap = cms.string(''),
    noise_MIP = cms.double(0.01),
    scaleByDose = cms.bool(False),
    scaleByDoseAlgo = cms.uint32(0)
)

So, technically, it brings back the proper config to mixing step. It should not effect HLT step. As hard-cloning configs are flying around as you also comment in [here](https://github.com/cms-sw/cmssw/pull/39450/files#r975724885), it would be better to converge on how to process for the moment, to allow 2026 menu to run at the proper step. So that we can run validation sequence on it without inventing new things.

This issue was mentioned in [here (presented in one of HLT upgrade meeting)](https://docs.google.com/document/d/1DU5MzzIeyGyXYbzTtfreCnsZ9FAmbI6_daFy0vklTos/edit#heading=h.wy1sntzfxqgd)

**Comment by missirol** [comment]
@srimanob , thanks. I think I understand better now.

So, the HGCal changes are technical, and needed to run the HLT Phase-2 menu in `step2`.

> This is not good for mixing process.

Clearly not. The HLT Phase-2 menu still contains many modules (producers, PSets, etc) named like they are in other steps (e.g. RECO), and this can lead to confusion, and unintended overwrites. This should change, but I cannot comment on when and how, as it was being handled by the HLT-Upgrade group. I'm actually surprised that the tests passed just with the few modifications in this PR.

> So that we can run validation sequence on it without inventing new things.

As you can see in the latest outputs of this PR, we don't really gain much information on the quality of the HLT Phase-2 reconstruction out-of-the-box. I think updates to the relevant DQM and Validation sequences will likely be needed (not in this PR, unless it is obvious how to do it).

In summary, I think this PR could go as it is, but running the current HLT Phase-2 in `step2` brings some risk (because of the whole renaming problem).

**Comment by srimanob** [comment]
OK, I come with the proposal.

I propose to
- update necessary psets, i.e. HGCAL psets which will not affect Phase-2 HLT results
- add a new workflow, i.e. `36434.76` which will be identical with `36434.0` except running `relval2026` menu. 

So, we don't touch the baseline workflow at all (so no risk), and we have an additional workflow for HLT upgrade to develop HLT monitoring for Phase-2 (without something fancy on top of `.75`).

Note that, I am trying to add back some HLT plots in validation sequence, but not success yet.

**Comment by srimanob** [comment]
test parameters:
- workflow = 39434.76
- relvals_opt = --what cleanedupgrade,standard,highstats,pileup,generator,extendedgen,production,ged,machine,premix

**Comment by srimanob** [comment]
@cmsbuild please test

**Comment by srimanob** [comment]
Example of HLT Validation from this PR: https://tinyurl.com/2kj6bdub
Screen Shot 2565-09-22 at 13 36 55

**Comment by missirol** [comment]
@srimanob 

Looks like we are adding thousands of empty DQM histograms in the wfs in question. The reason, at least in part, is that most of the HLT offline DQM hard-codes the names of Run-3 HLT collections and/or Paths. It's up to DQM to decide, but it looks like a more careful configuration would be needed.

**Comment by srimanob** [comment]
Hi @missirol 

Thanks. I noticed that too. Since it effects the phase2 validation sequence, so it adds more empty histograms.
I think when we agree on what to be done, I will remove DQM part and it can be a follow PR if I can find the way to limit them in general workflows.

**Comment by missirol** [comment]
> I think when we agree on what to be done, I will remove DQM part and it can be a follow PR if I can find the way to limit them in general workflows.

Regarding the scope of this PR, I thought the plan in https://github.com/cms-sw/cmssw/pull/39450#issuecomment-1253970416 was sensible (without the DQM updates that add many empty histograms).

Of course, the real development will be to (1) update the HLT Phase-2 to avoid any naming clashes with RECO or other steps, and any duplication of modules after this renaming, and (2) add monitoring for the collections in the HLT Phase-2 menu to spot issues like the one in #39323.

**Comment by srimanob** [comment]
Replace with the new PR https://github.com/cms-sw/cmssw/pull/39733
